Flux led config flow

icemanch opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Breaking change

Flux Led has fully transitioned to configuration via UI. YAML configuration will be removed in a future release. Existing YAML configuration is imported automatically and can now safely be removed from your configuration files.


## Proposed change
<!--
This is the first of three changes to update Flux_LED.  This change add config flow to the integration so that it can be used in the UI.
-->
- Add a config flow to flux_led

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #49130
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19566

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

icemanch committed:
Base update to add config_flow to Flux_LED

icemanch committed:
fixed config flow errors on integration removal

icemanch commented:
@emontnemery here is the first part of the Flux_LED update.  This is adding the config flow.

emontnemery commented:
Please add some basic tests for the config flow, it's mandatory.

icemanch committed:
Added testing

icemanch committed:
Exclude light and init from coverage

icemanch commented:
It looks like djtimca had tests setup in his PR so I pulled them over.  I'll have to do some reading to learn how to add testing for the other parts of this integration.  For now though is it enough to just have the config flow tested?   I did add the other files to .coveragrc

milanmeu commented on homeassistant/components/flux_led/__init__.py:
```diff
@@ -1 +1,110 @@
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["light"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Flux LED/MagicLight component."""
+    hass.data.setdefault(DOMAIN, {})
+
+    return True
```
`hass.data[DOMAIN]` is not used and `async_setup` is no longer required.
```suggestion
```

milanmeu commented on homeassistant/components/flux_led/const.py:
```diff
@@ -0,0 +1,24 @@
+
+API = "flux_api"
+
+ATTR_IDENTIFIERS = "identifiers"
+ATTR_MANUFACTURER = "manufacturer"
+ATTR_MODEL = "model"
```
Use the constants from `homeassistant/const.py` instead of adding integration constants for `ATTR_*`.

milanmeu reviewed (commented):
> For now though is it enough to just have the config flow tested?

Yes

icemanch committed:
Update homeassistant/components/flux_led/__init__.py

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

icemanch committed:
Update homeassistant/components/flux_led/config_flow.py

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

icemanch committed:
Update homeassistant/components/flux_led/__init__.py

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

icemanch committed:
Clean up of __init__

icemanch committed:
Changing const.py to import from HA

icemanch committed:
Remove ASYNC_SETUP from testing

icemanch commented:
Do I need to pass all of the pylint test?

emontnemery commented:
It's OK to disable pylint warnings or errors if they're overzealous.
This error should be addressed though:
```
Warning: homeassistant/components/flux_led/config_flow.py:39:4: W0102: Dangerous default value {} as argument (dangerous-default-value)
```

Just make `data` default to `None`, and set it to `{}` in the function instead.

There are many good explanations why mutable default arguments are bad, for example: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments.

emontnemery commented on homeassistant/components/flux_led/config_flow.py:
```diff
@@ -0,0 +1,234 @@
+                scanner = BulbScanner()
+                await self.hass.async_add_executor_job(scanner.scan)
+
+                for bulb in scanner.getBulbInfo():
+                    device_id = bulb["ipaddr"].replace(".", "_")
+                    if not devices.get(device_id, False):
+                        devices[device_id] = {
+                            CONF_NAME: bulb["ipaddr"],
+                            CONF_HOST: bulb["ipaddr"],
+                        }
```
Instead of duplicating code from the `FluxLedList` class here, can we use the `FluxLedList` class here instead of using `BulbScanner` directly? Or move this to a helper function which can be used both by the config flow and by `FluxLedList`?

emontnemery commented on homeassistant/components/flux_led/config_flow.py:
```diff
@@ -0,0 +1,234 @@
+        options = {
+            vol.Optional(
+                CONF_AUTOMATIC_ADD,
+                default=self._config_entry.options.get("global", {}).get(
+                    CONF_AUTOMATIC_ADD, self._config_entry.data[CONF_AUTOMATIC_ADD]
+                ),
```
We don't need to store the options in two places, we can instead pass options when creating the config entry in `async_step_user`, see https://github.com/home-assistant/core/pull/49912

emontnemery commented on homeassistant/components/flux_led/light.py:
```diff
@@ -21,24 +23,47 @@
-
 MODE_RGB = "rgb"
 MODE_RGBW = "rgbw"
+MODE_RGBCW = "rgbcw"
+MODE_RGBWW = "rgbww"
```
Why do we add these in this PR?

icemanch replied:
This PR was part of a larger PR that has been cut up. I tried to cut it down to be just config-related and get it back to the original code as much as possible.  Getting it all the way back to the original and just adding the config-related code will take a lot of work.

If that's what's needed I can do that but, keep in mind that the current integration in production barely functions.  This code keeps it function until it can be completely be updated.

bdraco replied:
Let's keep this PR to just the config flow. It will go much faster if there is less to review.

  
Additional fixes should come in a followup.

icemanch replied:
> Let's keep this PR to just the config flow. It will go much faster if there is less to review.
> 
> Additional fixes should come in a followup.

To be clear.  Do I take this PR back to draft and reduce the code further so that it's just the config flow?

emontnemery replied:
The review is already well underway but has mostly focused on the config flow until now.

If you push a commit which reverts changes to `homeassistant/components/flux_led/light.py` which are not required by the config flow implementation, that makes the review much easier, and reverting that commit can be the starting point for the next PR 👍

emontnemery commented on homeassistant/components/flux_led/light.py:
```diff
@@ -139,79 +166,119 @@
+
+    await async_new_lights(entry.data[CONF_DEVICES])
+
+    async_dispatcher_connect(hass, SIGNAL_ADD_DEVICE, async_new_lights)
```
I think we need to use the `on_ unload` helper to make sure the subscription is cancelled if the config entry is un- or reloaded, for example after the user changes options. If not, `async_new_lights` will be called multiple times and meaning we will have multiple entities for the same light.

A test to verify this is currently broken (you get multiple copies of a light) but is fixed by unsubscribing would be good.

emontnemery commented on homeassistant/components/flux_led/light.py:
```diff
@@ -139,79 +166,119 @@
+            try:
+                bulb = await hass.async_add_executor_job(WifiLedBulb, host)
+            except BrokenPipeError as error:
+                raise PlatformNotReady(error) from error
```
What happens if we end up here?

Also, we don't want to raise when handling `SIGNAL_ADD_DEVICE`?

emontnemery reviewed (commented):
Looks good overall, a few comments though.

bdraco commented on homeassistant/components/flux_led/config_flow.py:
```diff
@@ -0,0 +1,234 @@
+        """Get the options flow for the Flux LED component."""
+        return OptionsFlow(config_entry)
+
+    async def async_step_import(self, data: dict = {}):
```
Ideally we create a separate config entry for each host.  This avoids the need to build retry logic to reconnect to a device if it's offline during startup as we can raise ConfigEntryNotReady and auto retry later

icemanch replied:
My concern is that the only way to fire the scanner currently is through broadcasting.  It doesn't target specific IP's yet (future update).  I'll double-check on if it reports the MAC/Model directly as those would be ideal to use as the IDs.

bdraco replied:
If you send the packet directly to the ip, does it respond?

bdraco replied:
```python
diff --git a/flux_led/__main__.py b/flux_led/__main__.py
index 98c6488..cb4c2d8 100755
--- a/flux_led/__main__.py
+++ b/flux_led/__main__.py
@@ -1291,7 +1291,7 @@ class BulbScanner:
             if time.time() > quit_time:
                 break
             # send out a broadcast query
-            sock.sendto(msg, ("<broadcast>", DISCOVERY_PORT))
+            sock.sendto(msg, ("192.168.214.166", DISCOVERY_PORT))
 
             # inner loop waiting for responses
             while True:
```

```
bdraco@travelmacjnk-4 flux_led % python3 -m flux_led -sSti                   
18B905C5239B [192.168.214.166] ON  [Color: (0, 255, 0) White: 0 raw state: 129,51,35,97,35,1,0,255,0,0,9,0,240,84,]
  Timer #1: Unset
  Timer #2: Unset
  Timer #3: Unset
  Timer #4: Unset
  Timer #5: Unset
  Timer #6: Unset
```

Looks like they do respond

icemanch replied:
> If you send the packet directly to the ip, does it respond?

Yes, they respond.  I just need to update flux_led to make it work and didn't have time. I was going to update it after this PR is finished.

bdraco commented on homeassistant/components/flux_led/config_flow.py:
```diff
@@ -0,0 +1,234 @@
+                    if CONF_NAME in user_input
+                    else user_input[CONF_HOST]
+                )
+                device_id = user_input[CONF_HOST].replace(".", "_")
```
Please check https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

icemanch replied:
I noticed this and I was trying to work through it.  It's going to require changes to the Flux_LED API to get the Mac and the Model number from the devices.  Would you have a suggestion on how to get around this for now?

bdraco replied:
If you don't have the mac don't set a unique id.  You can always add it later

icemanch replied:
We have the Mac but only when using bulb_scanner.  My plan was to use the Model+MAC for the unique ID but, I was going to implement it once the scanner can probe specific IP addresses instead of using a broadcast.

Right, now the bulb scanner sends a broadcast out and then listens for responses.  I plan to change it so that it can also probe a specific address when an IP is passed to it.

bdraco commented on homeassistant/components/flux_led/manifest.json:
```diff
@@ -1,8 +1,9 @@
   "documentation": "https://www.home-assistant.io/integrations/flux_led",
   "requirements": ["flux_led==0.22"],
-  "codeowners": [],
+  "codeowners": ["@djtimca", "@icemanch"],
```
Is djtimca involved with this PR?

icemanch replied:
Most of the code in this pull request is belongs to him and I pulled it from a different Pull Request that had gone stale.  Which probably went stale because of me trying to update the Flux_LED API.

emontnemery replied:
We define code ownership like this:
> The willingness of a contributor to try, at best effort, to maintain the integration. Providing the intention for handling issues, providing bug fixes, or other contributions to the integration one is listed on as a code owner.

Unless @djtimca has agreed to this responsibility, let's not add him now.

bdraco commented on homeassistant/components/flux_led/__init__.py:
```diff
@@ -1 +1,95 @@
+
+PLATFORMS = ["light"]
+
+
```
It looks like the id is the mac address 
```
 % python3 -m flux_led -sSti                 
  warn(RuntimeWarning(msg))
18B905C5239B [192.168.214.166] ON  [Color: (0, 255, 0) White: 0 raw state: 129,51,35,97,35,1,0,255,0,0,9,0,240,84,]
  Timer #1: Unset
  Timer #2: Unset
  Timer #3: Unset
  Timer #4: Unset
  Timer #5: Unset
  Timer #6: Unset

```

bdraco replied:
We could do something similar to `tplink`

https://github.com/home-assistant/core/blob/dev/homeassistant/components/tplink/__init__.py#L107

Scan for devices in `async_setup` and create discovery flows for each one
https://github.com/home-assistant/core/blob/718f8d8bf796e9bb7cbdc29a7f2d19d79e9f5927/homeassistant/components/tplink/__init__.py#L65

If the ip address changes we update the config entry.

This would also allow setting a unique id on the config entry which could be used for the device that would meet the unique id requirements

icemanch replied:
That's correct it is the MAC.

icemanch committed:
Apply suggestions from code review

Co-authored-by: Erik Montnemery <erik@montnemery.com>

bdraco commented:
I'll pickup a few of these devices for testing

bdraco commented on homeassistant/components/flux_led/config_flow.py:
```diff
@@ -0,0 +1,232 @@
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
```
The user may want to enter the up address manually here.

Examples in both Yeelight and tplink

bdraco committed:
Single config entry per device

bdraco commented:
@icemanch If you cherry pick this commit into your branch, it will provide the basis for a config flow per device

https://github.com/bdraco/home-assistant/commit/7afa05e40c4ddf30294dee9dd125c880d9dfe93b

bdraco commented:
I went ahead and pushed the change to a single config per device. Of course feel free to revert it if you only want to use parts of it.

bdraco commented on homeassistant/components/flux_led/manifest.json:
```diff
@@ -1,8 +1,25 @@
   "requirements": ["flux_led==0.22"],
-  "codeowners": [],
-  "iot_class": "local_polling"
+  "codeowners": ["@djtimca", "@icemanch"],
```
```suggestion
  "codeowners": ["@icemanch"],
```

icemanch commented:
> I realize what I was asking was a big ask so I went ahead and pushed the change to a single config per device. Of course feel free to revert it if you only want to use parts of it. 

I'm super tied up at work.  I'm hoping to get back on this tomorrow evening or the weekend. I don't understand the confignflow very well so any help is greatly appreciated.  It seems like everyone says to look at other integrations to implement config flow but that doesn't really help in improving the understanding.  Is there any documentation anywhere on config flow?   

Honestly,  I'm basically hacking this code together.   Once I get the config flow done I can get into the stuff im actually familiar with.

bdraco commented:
> Is there any documentation anywhere on config flow?

https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#defining-your-config-flow

https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#testing-your-config-flow

bdraco commented:
Also feel free to reach out to me on Discard as `bdraco#6479`

bdraco commented on homeassistant/components/flux_led/__init__.py:
```diff
@@ -1 +1,201 @@
+
+PLATFORMS = ["light"]
+DISCOVERY_INTERVAL = timedelta(minutes=15)
+REQUEST_REFRESH_DELAY = 0.35
```
```suggestion
REQUEST_REFRESH_DELAY = 0.5
```
These appear to be slow to change state sometimes

bdraco replied:
sometimes it takes event longer so maybe not worth it since it self corrects every 5s

bdraco commented:
I have a few more devices that use this protocol showing up tomorrow.  I'll do some more testing

bdraco committed:
tests for config flow

bdraco committed:
coverage for options flow

bdraco committed:
Update homeassistant/components/flux_led/manifest.json

bdraco committed:
hassfest

bdraco commented on .coveragerc:
```diff
@@ -342,6 +342,8 @@ omit =
     homeassistant/components/flunearyou/sensor.py
     homeassistant/components/flux_led/light.py
+    homeassistant/components/flux_led/__init__.py
+    homeassistant/components/flux_led/const.py
```
```suggestion
```

bdraco committed:
Update .coveragerc

bdraco committed:
light coverage

bdraco committed:
coverage

bdraco committed:
cleanup

bdraco committed:
cleanup

bdraco committed:
update tests

bdraco committed:
test coordinator is updating

bdraco committed:
coverage

bdraco committed:
coverage

bdraco committed:
tweak tests

bdraco committed:
additional coverage

bdraco committed:
coverage

bdraco committed:
discovery tests

bdraco committed:
typing

bdraco committed:
remove unreachable

bdraco committed:
yaml migration test

bdraco committed:
fix registration

bdraco committed:
fix half implemented service

bdraco committed:
fix values

bdraco committed:
fix services.yaml

bdraco committed:
fix spurious error

bdraco committed:
fix selector

bdraco commented:
@icemanch I cleaned up things a bit more and added the missing tests.

It should be in a good place to make adjustments to fix your RGBWW/RGBCW lights.

bdraco committed:
enforce typing

bdraco committed:
hassfest

bdraco committed:
strict typing

icemanch commented:
> @icemanch I cleaned up things a bit more and added the missing tests.
> 
> It should be in a good place to make adjustments to fix your RGBWW/RGBCW lights.

Wow! You pretty much rewrote most of the code in 24 hours.  Testing it and it works really well.  A couple of questions.

1. Why did it have to go to a single entry?  I'm assuming it has to do with how you've made it set the brightness mode and the effects list. 
2. I had made several updates and tired to have a PR accepted but, @emontnemery thought was too big and request to have the scope narrowed and the updates done in stages.  Has this gotten too big again since it's more than the config flow @emontnemery ?
3. Can you help me understand the DHCP portion.  Is it filtering based on the MAC and model of the device?


Thanks for going through this.  I really was starting to spin my wheels.  I'm not sure how others have learned the config flow but I'm definitely struggling with it.

bdraco commented:
> > @icemanch I cleaned up things a bit more and added the missing tests.
> > It should be in a good place to make adjustments to fix your RGBWW/RGBCW lights.
> 
> Wow! You pretty much rewrote most of the code in 24 hours. Testing it and it works really well. A couple of questions.
> 
> 1. Why did it have to go to a single entry?  I'm assuming it has to do with how you've made it set the brightness mode and the effects list.

Since there was a custom effect list per device, it seemed too complex to try to manage that with a single config entry.

Also it allows us to use the built-in retry logic when the device is offline or the ip changes.

Finally, we don't have to worry about the device registry or entity registry since removing the config entry will clean it up automatically.

> 2. I had made several updates and tired to have a PR accepted but, @emontnemery thought was too big and request to have the scope narrowed and the updates done in stages.  Has this gotten too big again since it's more than the config flow @emontnemery ?

Its a bit larger than I would have liked, but the integration looks like it hasn't had a code owner in a while so it needed a lot more love than I was expecting when I really got into the cleanup.  Its got 100% coverage now so it should be lower risk to merge the changes now.

> 3. Can you help me understand the DHCP portion.  Is it filtering based on the MAC and model of the device?

https://developers.home-assistant.io/docs/creating_integration_manifest/#dhcp

> 
> Thanks for going through this. I really was starting to spin my wheels. I'm not sure how others have learned the config flow but I'm definitely struggling with it.

bdraco commented on homeassistant/components/flux_led/__init__.py:
```diff
@@ -1 +1,206 @@
+        )
+
+
+@callback
+def async_import_from_yaml(
+    hass: HomeAssistant,
+    config: ConfigType,
+    discovered_mac_by_host: dict[str, str],
+) -> None:
```
This isn't ideal because the platform schema is so complex. I think it would be better to do in an `async_setup_platform` in `light.py`, but to make that happen `discovered_mac_by_host` or discovered hosts need to be saved from `async_setup`

bdraco replied:
I'll refactor this a bit after the additional test devices show up

icemanch replied:
Tell me what you're looking for and I can test.

bdraco replied:
I don't have any of the RGBW, RGBWW, or RGBCW devices.

icemanch replied:
I'll test them right now.  

> Since there was a custom effect list per device, it seemed too complex to try to manage that with a single config entry.

The effect are set handled in the front end.  They don't really fit in the configuration as they are something that should be changed whenever you feel like it. 

![image](https://user-images.githubusercontent.com/10890432/135545932-95aca9f1-17ef-469b-9c57-81cf670941f4.png)

bdraco replied:
> I'll test them right now.  
> 
> 
> 
> > Since there was a custom effect list per device, it seemed too complex to try to manage that with a single config entry.
> 
> 
> 
> The effect are set handled in the front end.  They don't really fit in the configuration as they are something that should be changed whenever you feel like it. 
> 
> 
I agree, but I think we should try to avoid breaking users who are using them in that manner at the same time as it's converted to a config flow.  In the future, we could drop it and tell them to use the service.

icemanch replied:
I guess I never thought anyone would use it in the configuration since it's on the LIGHT card.  



> Remaining things to be tested:
> 
> * [ ]  RGBW, RGBWW, or RGBCW devices (I only have RGB)
> * [ ]  A complex yaml config import

I tested RGB, RGBW, and RGBWW.  RGBCW isn't a thing so I didn't test it.  

None of it works well but, it doesn't throw errors or break anything.  Pretty much works the way it did before config flow which was expected.  The bulbs turn on and off. You can set colors.  Color brightness can't be adjusted and neither can the master brightness. I have code ready to fix these but, I need to root around since some things have been moved around. 

I think we should get this PR accepted before adding anything else.  It will fix the issue most users are having.  Which is that the integration doesn't work at all.  At least with this, a user can turn it on and off. 

What do you think?

bdraco replied:
There is something off with this bulb

`129,53,35,97,23,5,255,0,0,0,8,0,240,77,`

Its detecting it as RGBW but its not

icemanch replied:
> There is something off with this bulb
> 
> `129,53,35,97,23,5,255,0,0,0,8,0,240,77,`
> 
> Its detecting it as RGBW but its not

Update flux_led through PyPi and run it.  It's on version 2.3. It may still not work because of they type 35 which is going to be added in over the weekend.  It has better type detection but, I missed the type 35s.

bdraco replied:
updated to 2.3 and still no go

bdraco replied:
If I go into the config options and change it from `auto` to `rgb` it works

icemanch replied:
> If I go into the config options and change it from `auto` to `rgb` it works

Yea, that's what I did to test.

The final goal will be working auto detect.  The flux_led 2.2 would only report if it was RGB capable and how many white channels.  Flux_LED 2.3 actually returns the exact type.  It makes it much easier to auto detect in homeassistant.  

There's two problems.  The first is that current code for the integration does not detect very well and only recognizes the 2.2 method of reporting.  The second problem is that your light is a type 35 which hasn't been added to the list of bulbs that use the new reporting in 2.3. It'll be fixed this weekend since it's a current issue.

bdraco replied:
The bulb says RGBCW on the side of it 🤷

bdraco replied:
One of these showed up and it it works in RGBCW mode

https://www.amazon.com/gp/product/B01DY56N8U/ref=ppx_yo_dt_b_asin_title_o07_s00?ie=UTF8&psc=1

bdraco replied:
```
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2145 B4E8422CB2B4 (b4:e8:42:2c:b2:b4) with raw_state=bytearray(b'\x813#a#\x01$\xff\x91\x00\t\x00\xf0\t') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2145 18B905C5239B (18:b9:05:c5:23:9b) with raw_state=bytearray(b'\x813#a#\x10\x00\x00\xff\x00\t\x00\xf0c') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ200 8CCE4E03EE81 (8c:ce:4e:03:ee:81) with raw_state=bytearray(b'\x81%#a\x00\xff\xff\x00\x00\x00\x02\x00\x00*') is rgbww
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2145 B4E8424401EE (b4:e8:42:44:01:ee) with raw_state=bytearray(b'\x813#a#\x10\x00\x00\xff\x00\t\x00\xf0c') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2145 B4E84214155A (b4:e8:42:14:15:5a) with raw_state=bytearray(b'\x815#a\x17\x05\x00\xff\x00\x00\x08\x00\xf0M') is rgbww
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2146 249494176B09 (24:94:94:17:6b:09) with raw_state=bytearray(b'\x813#a#\x10\x00?\xff\x00\n\x00\xf0\xa3') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2146 249494861B97 (24:94:94:86:1b:97) with raw_state=bytearray(b'\x813#a#\x10\xff\x00\x00\x00\n\x00\xf0d') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2146 24949416B10D (24:94:94:16:b1:0d) with raw_state=bytearray(b'\x813#a#\x10\x00\x00\xff\x00\n\x00\xf0d') is rgb
```

icemanch replied:
> One of these showed up and it it works in RGBCW mode
> 
> https://www.amazon.com/gp/product/B01DY56N8U/ref=ppx_yo_dt_b_asin_title_o07_s00?ie=UTF8&psc=1
> ```
> % python3 -m flux_led 192.168.214.125 --info
> Unknown ID [192.168.214.125] ON  [Color: (0, 0, 255) White: 0 raw state: 129,37,35,97,0,255,0,0,255,0,2,0,0,42,]
> ```

Yes and you can change the mode.   Those are what use to test the updated detection l.  In the app if you long press the light there's an option to change type.

icemanch replied:
> The bulb says RGBCW on the side of it 🤷 

The type is 35 which is an RGBWW bulb but Flux_LED doesn't have a definition for 35 yet.

The other controllers you have are defined in Flux_LED.

bdraco replied:
I've tested with everything else and it appears to be working as expected.

bdraco replied:
Based on this issue https://github.com/Danielhiversen/flux_led/issues/79 it looks like the same problem.  I'm comfortable its not a regression.  We can work though that after this merges

bdraco committed:
improve robustness of yaml import

bdraco commented on homeassistant/components/flux_led/__init__.py:
```diff
@@ -1 +1,142 @@
+async def async_discover_devices(
+    hass: HomeAssistant, timeout: int
+) -> list[dict[str, str]]:
+    """Discover ledned devices."""
```
```suggestion
    """Discover flux led devices."""
```

bdraco committed:
Update homeassistant/components/flux_led/__init__.py

bdraco committed:
log detected mode

bdraco committed:
logging

bdraco commented on homeassistant/components/flux_led/light.py:
```diff
@@ -92,39 +128,36 @@
+EFFECT_ID_NAME: Final = {v: k for k, v in EFFECT_MAP.items()}
+EFFECT_CUSTOM_CODE: Final = 0x60
+
+WHITE_MODES: Final = {MODE_RGBW, MODE_RGBWW, MODE_RGBCW}
```
@icemanch Please check this logic to make sure it makes sense

icemanch replied:
Trying to use RGBWW or RGBCW is going to add a lot of unnecessary code that's going to get overwritten.   I removed both as there really isn't a good way for Home Assistant to figure out if it is RGBWW capable.

bdraco commented on homeassistant/components/flux_led/light.py:
```diff
@@ -333,42 +415,62 @@ def turn_on(self, **kwargs):
 
         # handle RGBW mode
-        elif self._mode == MODE_RGBW:
+        if self._mode in WHITE_MODES:
```
Used here as well

icemanch replied:
This won't work.  RGBWW requires setting the 5th channel which will require calculating the brightness. CCT, and RGB values.

bdraco commented on homeassistant/components/flux_led/light.py:
```diff
@@ -333,42 +415,62 @@ def turn_on(self, **kwargs):
             self._bulb.setRgbw(0, 0, 0, w=brightness)
+            return
+
+        if white is None and self._mode in WHITE_MODES:
```
And here

icemanch replied:
Yes, this works.

bdraco commented:
We probably need something in the docs to tell the user to change the mode if auto detection gets it wrong

bdraco commented:
With all the magic numbers it's really difficult to quickly see the state.  It probably makes sense to add a property to flux_led to return the current state as a named tuple or data class (probably too new for all supported use cases)

icemanch commented:
> With all the magic numbers it's really difficult to quickly see the state.  It probably makes sense to add a property to flux_led to return the current raw_state as a named tuple or data class (probably too new for all supported use cases)

2.3 does that for recognized devices.   It sends back the current mode of the device if it's been recognized.  If it's not recognized or fails back to the original data.

bdraco commented:
Should we bump the dep to 2.3 in this pr? Are there any changes needed on then home assistant side?

bdraco commented:
It looks like the parts I wanted Erik to take a peek at are all going to be re-written here https://github.com/home-assistant/core/pull/52057

icemanch commented:
> Should we bump the dep to 2.3 in this pr? Are there any changes needed on then home assistant side?

No.  The integration isn't ready for it.  The next release will cover it.

icemanch commented:
> It looks like the parts I wanted Erik to take a peek at are all going to be re-written here #52057

Yes, next step is adding supported color modes.

icemanch committed:
Remove RGBWW and RGBCW.

bdraco commented:
Documentation added/updated for www.home-assistant.io is checked, but I couldn't find the docs PR for `next` branch at https://github.com/home-assistant/home-assistant.io

We add a link under `Link to documentation pull request:` in the template for any PRs that need a docs change.

bdraco commented:
Documentation added/updated for www.home-assistant.io is checked, but I couldn't find the docs PR for `next` branch at https://github.com/home-assistant/home-assistant.io

We add a link under `Link to documentation pull request:` in the template for any PRs that need a docs change.

bdraco reviewed (approved):
I ran though this again and everything looks good with your latest push.

Once we have the docs PR opened, this should be good to merge.

bdraco reviewed (approved):
I ran though this again and everything looks good with your latest push.

Once we have the docs PR opened, this should be good to merge.

bdraco commented:
The 9w version showed up today https://www.amazon.com/gp/product/B08NT4KZ4Z/ref=ppx_yo_dt_b_asin_title_o03_s00?ie=UTF8&psc=1

It worked out of box with your latest push

`Unknown ID [192.168.106.235] ON  [Warm White: 0% raw state: 129,53,35,97,1,16,163,163,163,0,7,0,240,43,]`

`2021-10-01 17:56:26 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2104 2462AB6BD660 (24:62:ab:6b:d6:60) with raw_state=bytearray(b'\x815#a\x01\x05\x00\xff\x00\x00\x07\x00\xf06') rgbwcapable=False is rgb`

bdraco commented:
I also picked up some of the smart sockets
```
% python3 -m flux_led 192.168.106.147 -i
/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py:127: RuntimeWarning: 'flux_led.__main__' found in sys.modules after import of package 'flux_led', but prior to execution of 'flux_led.__main__'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/homebrew/lib/python3.9/site-packages/flux_led/__main__.py", line 2095, in <module>
    main()
  File "/opt/homebrew/lib/python3.9/site-packages/flux_led/__main__.py", line 2072, in main
    print("{} [{}] {}".format(info["id"], info["ipaddr"], bulb))
  File "/opt/homebrew/lib/python3.9/site-packages/flux_led/__main__.py", line 766, in __str__
    pattern = rx[3]
TypeError: 'NoneType' object is not subscriptable
```

bdraco commented:
Looks like the lib will need some updating before we could support those

```
>>> import flux_led
>>> socket = flux_led.WifiLedBulb('192.168.106.147')
['>', bytearray(b'\x81\x8a\x8b\x96')]
['<', b'\x81\x97']
['>', bytearray(b'\x81\x8a\x8b\x96')]
['<', b'\x81\x97$$\x00\x00\x00\x00\x00\x00\x02\x00\x00b']
['>', bytearray(b'\x81\x8a\x8b\x96')]
['<', b'\x81\x97$$\x00\x00\x00\x00\x00\x00\x02\x00\x00b']
['>', bytearray(b'\x81\x8a\x8b\x96')]
['<', b'\x81\x97$$\x00\x00\x00\x00\x00\x00\x02\x00\x00b']
>>> print(socket)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bdraco/flux_led/flux_led/__main__.py", line 766, in __str__
    pattern = rx[3]
TypeError: 'NoneType' object is not subscriptable
```

bdraco commented:
`Unknown ID [192.168.106.147] OFF  [Switch raw state: 129,151,36,36,0,0,0,0,0,0,2,0,0,98,]`

```
diff --git a/flux_led/__main__.py b/flux_led/__main__.py
index 98c6488..ba63b23 100755
--- a/flux_led/__main__.py
+++ b/flux_led/__main__.py
@@ -616,6 +616,8 @@ class WifiLedBulb:
                 mode = "color"
         elif pattern_code == 0x60:
             mode = "custom"
+        elif pattern_code == 0x24:
+            mode = "switch"
         elif pattern_code == 0x62:
             mode = "music"
         elif pattern_code == 0x41:
@@ -804,6 +806,8 @@ class WifiLedBulb:
             mode_str = BuiltInTimer.valtostr(pattern)
         elif mode == "music":
             mode_str = "Music"
+        elif mode == "switch":
+            mode_str = "Switch"
         else:
             mode_str = "Unknown mode 0x{:x}".format(pattern)
         mode_str += " raw state: "
```

bdraco committed:
tweak refresh delay as these devices can take > 0.6s to update internal state

bdraco commented:
I found the original docs PR (https://github.com/home-assistant/home-assistant.io/pull/17191) that has gone stale.

As soon as a new one is opened and linked this should be good to merge 👍

icemanch commented:
> I found the original docs PR ([home-assistant/home-assistant.io#17191](https://github.com/home-assistant/home-assistant.io/pull/17191)) that has gone stale.
> 
> As soon as a new one is opened and linked this should be good to merge 👍

So I need to open a PR for the documentation and link it to this PR correct?  Then I need to update the documentation to support the changes?

bdraco commented:
> So I need to open a PR for the documentation and link it to this PR correct? Then I need to update the documentation to support the changes?

Yes

https://github.com/home-assistant/home-assistant.io/edit/next/source/_integrations/flux_led.markdown

You could scrape most of what was in https://github.com/home-assistant/home-assistant.io/pull/17191 as its almost all still relevant.

You can also replace most of the instructions on how setup with `{% include integrations/config_flow.md %}` as it will include the template for integrations with a config flow.
